pkg/p2p/peer/peer.go

```diff
@@ -180,18 +180,31 @@ func (p *Reader) Accept() error {
 	return nil
 }
 
-// Serve utilizes two different methods for writing to the open connection
-func (w *Writer) Serve(writeQueueChan <-chan bytes.Buffer, exitChan chan struct{}) {
-	defer w.onDisconnect()
+// Create two-way communication with a peer. This function serves to split up
+// the passed context and allowing either goroutine to cancel the other
+// in case of errors.
+func Create(ctx context.Context, reader *Reader, writer *Writer, writeQueueChan <-chan bytes.Buffer) {
```

Comment:
The context should be passed to the functions doing the work (in this case `reader.ReadLoop` and `writer.Serve`), and not state-wide to a struct instance or a closure. Also, I don't like tightly coupled goroutines and I would advocate for communicating results only to the caller rather than influence each other with a direct cancelation.

Something like the following:

```golang
func Create(ctx context.Context, reader *Reader, writer *Writer, writeQueueChan <-chan bytes.Buffer) {
    ...
    ctx, cancel := context.WithCancel(ctx)
    defer cancel()

    errChan := make(chan error, 1)
    
    go reader.ReadLoop(ctx, writeQueueChan, errChan)
    go writer.WriteLoop(ctx, writeQueueChan, errChan)

    err := <-errChan
    if !err.Is(ErrDisconnect) {
        // log it
    }
}
``` 
if you want to bubble up any error to the server in the `OnAccept` you can do it easily by passing the `errChan` similarly to what you do with `writeQueueChan`. In this case, the `ErrDisconnect` should be known project-wise